File: leet-code/147. Insertion Sort List/InsertionSortList147_3.java
package com.example;

public class InsertionSortList147_3 {
    public ListNode insertionSortList(ListNode head) {
        ListNode root = new ListNode();
        ListNode prev = root;
        while (head != null) {
            while (prev.next != null && prev.next.val < head.val) {
                prev = prev.next;
            }
            ListNode prevNext = prev.next;
            ListNode headNext = head.next;
            prev.next = head;
            head.next = prevNext;
            head = headNext;

            if (head != null && prev.val > head.val) {
                prev = root;
            }
        }
        return root.next;
    }
}
// 2 4 1 3 5
// p h
//     h

File: leet-code/147. Insertion Sort List/InsertionSortList147_3.kt
package com.example

class InsertionSortList147_3 {
    fun insertionSortList(head: ListNode?): ListNode? {
        var headNode = head
        val sortedNode = ListNode()
        while (headNode != null) {
            var cur = sortedNode
            while (cur.next != null && cur.next.`val` < headNode.`val`) {
                cur = cur.next
            }
            val temp = headNode
            headNode = headNode.next
            temp.next = null
            val curNext = cur.next
            cur.next = temp
            cur.next.next = curNext
        }
        return sortedNode.next
    }
}
